build_runner/lib/src/entrypoint/doctor.dart

```diff
@@ -0,0 +1,123 @@
+import 'package:logging/logging.dart';
+
+import '../logging/std_io_logging.dart';
+import '../package_graph/build_config_overrides.dart';
+import 'base_command.dart';
+
+/// A command that validates the build environment.
+class DoctorCommand extends BuildRunnerCommand {
+  @override
+  String get name => 'doctor';
+
+  @override
+  bool get hidden => true;
+
+  @override
+  String get description => 'Check for misconfiguration of the build.';
+
+  @override
+  Future<int> run() async {
+    final options = readOptions();
+    final verbose = options.verbose ?? false;
+    Logger.root.level = verbose ? Level.ALL : Level.INFO;
+    final logSubscription =
+        Logger.root.onRecord.listen(stdIOLogListener(verbose: verbose));
+
+    final config = await _loadBuilderDefinitions();
+
+    var isOk = true;
+    for (final builderApplication in builderApplications) {
+      final builderOk = _checkBuildExtensions(builderApplication, config);
+      isOk = isOk && builderOk;
+    }
+
+    if (isOk) {
+      logger.info('No problems found!\n');
+    }
+    await logSubscription.cancel();
+    return isOk ? ExitCode.success.code : ExitCode.config.code;
+  }
+
+  Future<Map<String, BuilderDefinition>> _loadBuilderDefinitions() async {
+    final packageGraph = PackageGraph.forThisPackage();
+    final buildConfigOverrides =
+        await findBuildConfigOverrides(packageGraph, null);
+    Future<BuildConfig> _packageBuildConfig(PackageNode package) async {
+      if (buildConfigOverrides.containsKey(package.name)) {
+        return buildConfigOverrides[package.name];
+      }
+      try {
+        return await BuildConfig.fromBuildConfigDir(package.name,
+            package.dependencies.map((n) => n.name), package.path);
+      } on ArgumentError catch (e) {
+        logger.severe(
+            'Failed to parse a `build.yaml` file for ${package.name}', e);
+        return BuildConfig.useDefault(
+            package.name, package.dependencies.map((n) => n.name));
+      }
+    }
+
+    final allConfig = await Future.wait(
+        packageGraph.allPackages.values.map(_packageBuildConfig));
+    final allBuilders = <String, BuilderDefinition>{};
+    for (final config in allConfig) {
+      allBuilders.addAll(config.builderDefinitions);
+    }
+    return allBuilders;
+  }
+
+  /// Returns true of [builderApplication] has sane build extension
+  /// configuration.
+  ///
+  /// If there are any problems they will be logged and `false` returned.
+  bool _checkBuildExtensions(BuilderApplication builderApplication,
+      Map<String, BuilderDefinition> config) {
+    var phases = builderApplication.buildPhaseFactories
+        .map((f) => f(PackageNode(null, null, null, isRoot: true),
+            BuilderOptions.empty, InputSet.anything, InputSet.anything, true))
+        .whereType<InBuildPhase>()
+        .toList();
+    if (phases.isEmpty) return true;
+    if (!config.containsKey(builderApplication.builderKey)) return false;
```

Comment:
why is this check necessary?

Reply:
The `BuilderApplications` we have are from the build script snapshot, and there could be slight variance in the builders that are in the loaded config.

There's going to be some weirdness around shapshots and stuff that we will want to work out here - we should iterated on the doctor command for a while before we expose it in the help output.